inspirehep/modules/orcid/cache.py

```diff
@@ -31,69 +31,114 @@
 from .exceptions import EmptyPutcodeError
 
 
+# Redis client as module-level singleton.
+_redis_client = None
+
+
+def _init_redis_client():
+    """
+    Note: the `_redis_client` module-level var cannot be initialized at module-load
+    time because accessing flask.current_app raises the exception:
+    "RuntimeError: Working outside of application context"
+    """
+    url = app.config.get('CACHE_REDIS_URL')
+    global _redis_client
+    _redis_client = StrictRedis.from_url(url)
+
+
 class OrcidCache(object):
     def __init__(self, orcid):
-        redis_url = app.config.get('CACHE_REDIS_URL')
-        self.redis = StrictRedis.from_url(redis_url)
+        """
+        Orcid cached data.
+
+        Args:
+            orcid (string): orcid identifier.
+        """
         self.orcid = orcid
+        self._cached_hash_value = None
+        self._new_hash_value = None
+        if not _redis_client:
+            _init_redis_client()
+        self.redis = _redis_client
 
-    def _get_key(self, recid):
+    def get_key(self, recid):
         """Return the string 'orcidcache:``orcid_value``:``recid``'"""
         return 'orcidcache:{}:{}'.format(self.orcid, recid)
 
-    def write_record_data(self, recid, putcode, hash_value=None):
+    def write_work_putcode(self, recid, putcode, inspire_record=None):
         """
         Write the putcode and the hash for the given (orcid, recid).
 
         Args:
-            recid(int): inspire record's id.
-            putcode(string): the putcode used to push the record to ORCID.
-            hash_value(Optional[string]): hashed ORCID record content.
+            recid (string): record identifier.
+            putcode (string): the putcode used to push the record to ORCID.
+            inspire_record (InspireRecord): InspireRecord instance. If provided,
+             the hash for the record content is re-computed.
         """
         if not putcode:
             raise EmptyPutcodeError
 
-        key = self._get_key(recid)
-        value = {'putcode': putcode}
-        if hash_value:
-            value['hash'] = hash_value
-        self.redis.hmset(key, value)
+        data = {'putcode': putcode}
+
+        if inspire_record:
+            if not self._new_hash_value:
+                self._new_hash_value = _OrcidHasher(inspire_record).compute_hash()
+            data['hash'] = self._new_hash_value
 
-    def read_record_data(self, recid):
+        key = self.get_key(recid)
+        self.redis.hmset(key, data)
+
+    def read_work_putcode(self, recid):
```

Comment:
this also gets the hash from the cache, which is not obvious from the name.

Reply:
See prev comment about the concept of hash being an internal implementation detail of the cache object. Thus this methods "reads and returns a putcode for a given (in the constructor in the last version) orcid, recid"